ghcide/src/Development/IDE/Plugin/Completions.hs

```diff
@@ -156,17 +156,37 @@ getCompletionsLSP ide plId
                 let clientCaps = clientCapabilities $ shakeExtras ide
                 config <- getCompletionsConfig plId
                 allCompletions <- liftIO $ getCompletions plId ideOpts cci' parsedMod bindMap pfix' clientCaps config moduleExports
-                pure $ InL (List allCompletions)
+                pure $ InL (List $ orderedCompletions allCompletions)
               _ -> return (InL $ List [])
           _ -> return (InL $ List [])
       _ -> return (InL $ List [])
 
+{- COMPLETION SORTING
+   We return an ordered set of completions (local -> nonlocal -> global).
+   Ordering is important because local/nonlocal are import aware, whereas
+   global are not and will always insert import statements, potentially redundant.
+
+   On the other hand the fuzzy sort algorithm doesn't always sort in the optimal way,
+   there is room for the LSP client to improve.
+
+   According to the LSP specification, if no sortText is provided, the label is used.
+   To allow the LSP client to reorder identifiers while preserving the relative ordering
+   of repeated occurrences we generate sortText values that include both the label and
+   an index denoting the relative order
```

Comment:
If I understand this right, given a list of completions like:
```
-- local
x
y
-- global
x
y
```
this will produce
```
x -- local
x -- global
y -- local
y -- global
```

Is that what we want? Maybe it would be better to put *all* the local ones first, i.e. have the sort text be just the index. The client can still filter on the label (or we can set `filterText` explicitly). Either way, would be good to have an example in this comment.

Reply:
This is what we want if the LSP client thinks that `x` is more relevant than `y`. The reason is that the fuzzy sort package that we use doesn't do an optimal job imho, so we want to allow the LSP client to improve on that as long as it keeps local before global

Reply:
I'll extend the comment later

Reply:
Okay, so we're basically adding a second measure of completion "goodness", which is lexicographically less relevant than the text? That makes sense to me.

(Tangent: this makes me wonder whether we can skip name-based sorting on the server at all? I know you did a bunch of work making that faster... the fastest thing is not to do it! Perhaps some clients don't sort and so it's helpful to have some sort order from our end?)

Reply:
Or, if it's just the local/nonlocal/global thing, we could just add 1/2/3 priority digits to the sort text when we construct them. The ordering here is a perhaps unnecessarily fine-grained. Not that it makes much difference.

Reply:
> (Tangent: this makes me wonder whether we can skip name-based sorting on the server at all? I know you did a bunch of work making that faster... the fastest thing is not to do it! Perhaps some clients don't sort and so it's helpful to have some sort order from our end?)

That was sorting the fuzzy matches by score - definitely still needed. We don't sort alphabetically otherwise

Reply:
> Or, if it's just the local/nonlocal/global thing, we could just add 1/2/3 priority digits to the sort text when we construct them. The ordering here is a perhaps unnecessarily fine-grained. Not that it makes much difference.

This would result in all the local completions first, regardless of how fuzzy, then the global. It would not be what we want. 

But we ought to make the fuzzy score visible to the LSP client in the sortText field, so that the sorting respects it.